### PR TITLE
fix(ci): stop main-push release runs cancelling each other; self-heal dropped app releases

### PR DESCRIPTION
**Pull request by jthrilly**
## Problem

New **interviewer** (and **architect**) beta releases stopped going out. No GitHub releases appear and the PWAs show no updates. `apps/interviewer` sits at `8.0.0-beta.2` and `apps/architect` at `8.0.0-beta.3`, but neither has a git tag or release — and interviewer has never released under its new lane at all.

## Root cause

Every push to `main` shared **one** concurrency group:

```yaml
concurrency:
  group: ci-${{ github.ref }}          # "ci-refs/heads/main" for ALL main pushes
  cancel-in-progress: false            # for push events
```

A merge queue is enabled, so PRs land on `main` in bursts (7 merges in ~10 min on 2026‑07‑07). GitHub allows only **one in‑progress + one pending** run per group and cancels the rest **while queued, with zero jobs scheduled**. The push that merged the *Release apps (beta)* PR #830 (interviewer `beta.1→beta.2`, architect `beta.2→beta.3`) was one of the evicted runs (`28833660789`, 0 jobs, `cancelled`), so `apps-release-detect` → `apps-release-interviewer`/`apps-release-architect` never ran.

`detect-app-release.sh` then made it **permanent**: it only fired on a `HEAD^ → HEAD` beta increment, so once the bump commit scrolled into history no later push could re-detect it. One cancelled run lost the release for good.

## Fix

1. **Per-commit concurrency for push-to-main** — main pushes no longer cancel each other; PR runs keep per-ref grouping + `cancel-in-progress`.
2. **Tag-driven, self-healing detection** — release a beta (`N > 0`) whose tag is missing. Still idempotent (existing tag → no-op); now a dropped release is caught up on the next push.

Adds `scripts/detect-app-release.test.mjs` (runs in the `test-scripts` job): covers normal release, idempotency, the **self-heal** regression (no `HEAD^` delta + missing tag → release), the `beta.0` seed, and non-beta versions.

## Recovery

Merging this **automatically recovers both stuck betas**: the merge's push run gets a per-commit group (can't be evicted), and the new detection sees `@codaco/architect@8.0.0-beta.3` and `@codaco/interviewer@8.0.0-beta.2` have no tags → releases both from current `main`. No manual re-run of the stale cancelled snapshot needed.

> [!NOTE]
> `apps-release-interviewer` deploys to Netlify prod using `NETLIFY_SITE_ID_INTERVIEWER` — confirm that secret is set, or that job will fail at the deploy step (the release/tag steps run after it).

## Verification

- `pnpm test:scripts` → 19/19 pass (5 new)
- `oxlint`, `oxfmt --check`, `actionlint`, `shellcheck` all clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved app release detection to use git tags, reducing false negatives and preventing duplicate release attempts.
  * Refined CI concurrency so push builds on different commits don’t cancel each other unexpectedly, while PR runs still supersede older ones.

* **Tests**
  * Added coverage for release detection scenarios, including new beta releases, already-tagged versions, non-beta versions, and edge cases around initial beta builds.

### CODE REVIEW

File: scripts/detect-app-release.test.mjs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  git(cwd, 'config', 'user.email', 'ci@example.com');
+  git(cwd, 'config', 'user.name', 'ci');
+
+  const writePkg = (v) =>
+    writeFileSync(
+      join(cwd, PKG_JSON),
+      `${JSON.stringify({ name: PKG_NAME, version: v, private: true }, null, 2)}\n`,
+    );
+
+  // First commit establishes HEAD^ so the script can read the previous version.
```
[the remark this answers is not in the document]
Comment (jthrilly) [reply scripts/detect-app-release.test.mjs]: Fixed in d77e94115 — reworded this comment to state the real intent (the script is tag-driven and ignores the previous commit; the setup exists to prove detection does not depend on a HEAD^ diff).
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: scripts/detect-app-release.test.mjs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  writePkg(previousVersion ?? version);
+  git(cwd, 'add', '.');
+  git(cwd, 'commit', '-qm', 'first');
+
+  writePkg(version);
+  git(cwd, 'add', '.');
+  // --allow-empty so the no-delta cases (previousVersion === version) still get
+  // a HEAD^, reproducing a bump whose commit has already scrolled into history.
```
[the remark this answers is not in the document]
Comment (jthrilly) [reply scripts/detect-app-release.test.mjs]: Fixed in d77e94115 — the `--allow-empty` comment no longer references HEAD^; it now just explains the second commit keeps the same tree so the no-delta case simulates a later push after the bump commit has scrolled past.
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/scripts/detect-app-release.sh [hunk 3]
```diff
@@ -2,46 +2,43 @@
...
+reason="not a releasable beta (version=$current)"
+if [[ "$current" =~ $beta_re ]] && [[ "${BASH_REMATCH[2]}" -gt 0 ]]; then
+  if git rev-parse -q --verify "refs/tags/$PKG_NAME@$current" >/dev/null; then
+    reason="tag $PKG_NAME@$current already exists"
+  else
     released=true
-    reason="beta increment $previous -> $current"
+    reason="beta $current has no release tag yet"
   fi
 fi
```
[the remark this answers is not in the document]
Comment (jthrilly) [reply .github/scripts/detect-app-release.sh:46]: Good catch — this race is a direct consequence of the new per-commit push concurrency (previously all main pushes serialised in one group, so two release runs for the same untagged version could never overlap). Fixed in 70af221e6: added a per-`<PKG_NAME>@<version>` concurrency group (`cancel-in-progress: false`) to `apps-release-architect`/`apps-release-interviewer` so runs targeting the same version serialise, plus a tag re-check guard (checkout now `fetch-tags: true`) that skips build/deploy/release when the tag already exists — so the serialised second run is a clean green no-op instead of a wasted prod deploy. Updated the detect-app-release.sh header to note its tag check is only a preflight and the authoritative guard lives on the release jobs.
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
